# `Memgraph`  

*  high-performance in-memory graph database  
*  faster access and efficient in-memory data processing

In [7]:
import mgclient
import pprint as pp

## Build a `Memgraph` Database

In [ ]:
# # Delete all nodes and relationships
# cursor.execute("MATCH (n) DETACH DELETE n")

### Connect to the Database

In [3]:
conn = mgclient.connect(host="localhost", port=7687)
conn.autocommit = True
cursor = conn.cursor()
cursor.execute("CALL mg.load_all();") # Load MAGE



In [ ]:
# Load only the specific module
cursor.execute("CALL mg.load('algo');")

### Inspect Database

Fetching a Single Document: To view a single document (which, in the context of a graph database, typically refers to a node or a relationship), you might execute a query like:



In [11]:
cursor.execute("MATCH (n) RETURN n LIMIT 1")
single_node = cursor.fetchone()
print(single_node)


(<mgclient.Node(id=964, labels={'Subject'}, properties={'name': 'SmallMolecule'}) at 0x111b6d9b0>,)


``` 
(<mgclient.Node(id=964, labels={'Subject'}, properties={'name': 'SmallMolecule'}) at 0x111b6d6e0>,)
```
- a tuple containing a single `mgclient.Node` object  

- `mgclient.Node`: This indicates that the object is a node in the Memgraph graph database.

- `id=964`: This is the unique identifier of the node within the Memgraph database. Each node has a unique ID.  

- `labels={'Subject'}`: This shows the labels associated with the node. In graph databases like Memgraph, labels are used to categorize nodes into different types. In this case, the node has a label of `'Subject'`, indicating its role or type within your graph schema.  

- `properties={'name': 'SmallMolecule'}`: This is a dictionary of the properties of the node. Properties are key-value pairs that store additional information about nodes. In this example, the node has one property named name, with the value 'SmallMolecule'. This could represent the name or title of the subject in your graph.  

- `at 0x111a77a20`: This is the memory address where this particular mgclient.Node object is stored in Python. It's not relevant for understanding the data in your graph, but it's a normal part of Python's representation of objects.

- `Tuple Representation`: The entire node object is enclosed in a tuple, as indicated by the parentheses ( and ). This is typical when fetching results from a database in Python; even if there's only one result, it's often returned as a single-item tuple.

In [13]:
node_data = cursor.fetchone()
if node_data is not None:
    node = node_data[0]  # Extract the node from the tuple
    node_id = node.id
    node_label = list(node.labels)[0]  # Assuming each node has only one label
    node_name = node.properties['name']
    print(f"Node ID: {node_id}, Label: {node_label}, Name: {node_name}")
else:
    print("No data returned from the query.")


No data returned from the query.


Exploring Relationships: To see how Subject and Object nodes are related, you could use a query like:

In [5]:
cursor.execute("MATCH (s:Subject)-[r:RELATION]->(o:Object) RETURN s, r, o LIMIT 5")
relationships = cursor.fetchall()
for rel in relationships:
    print(rel)


(<mgclient.Node(id=964, labels={'Subject'}, properties={'name': 'SmallMolecule'}) at 0x111b52cd0>, <mgclient.Relationship(start_id=964, end_id=965, type='RELATION', properties={'api': [{'bte': {'query_operation': {'input_separator': ',', 'method': 'post', 'params': None, 'path': '/query', 'path_params': None, 'request_body': None, 'server': 'https://automat.transltr.io/icees-kg/1.3', 'support_batch': True}}, 'name': 'Automat-icees-kg(Trapi v1.3.0)', 'smartapi': {'id': '44e7a1147ca8657f50af6bb25982762d', 'metadata': 'https://trapi-openapi.apps.renci.org/trapi/infores:automat-icees-kg/1.3.0', 'ui': 'https://smart-api.info/ui/44e7a1147ca8657f50af6bb25982762d'}, 'tags': ['translator', 'automat', 'trapi', 'bte-trapi'], 'x-translator': {'component': 'KP', 'team': ['Exposures Provider']}}, {'bte': {'query_operation': {'input_separator': ',', 'method': 'post', 'params': None, 'path': '/query', 'path_params': None, 'request_body': None, 'server': 'https://bte.transltr.io/v1', 'support_batch': T